File: python/has_cycle.py
class Solution(object):
    # O(n) time | O(1) space
    def hasCycle(self, head):
        """
        :type head: ListNode
        :rtype: bool
        """
        dummy = ListNode(0)
        dummy.next = head
        slow = dummy
        fast = head

        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next

            if slow == fast:
                return True
        
        return False


class ListNode(object):
    def __init__(self, val, next):
        self.val = val
        self.next = next
    
    def __repr__(self) -> str:
        res = str(self.val)
        if self.next:
            res += str(self.next)
        return res

head = ListNode(3, None)
head.next = ListNode(2, None)
head.next.next = ListNode(0, None)
head.next.next.next = ListNode(4, None)
head.next.next.next.next = head.next

solution = Solution()
print(solution.hasCycle(head))

File: src/main/java/june_2025/HasCycle.java
package june_2025;

public class HasCycle {

    public static void main(String[] args) {
        ListNode head = new ListNode(3);
        head.next = new ListNode(2);
        head.next.next = new ListNode(0);
        head.next.next.next = new ListNode(-4);
        head.next.next.next.next  = head.next;

        HasCycle hasCycle = new HasCycle();
        boolean result = hasCycle.hasCycle(head);
        System.out.println(result);
    }

    public boolean hasCycle(ListNode head) {
        ListNode dummy = new ListNode(0);
        dummy.next = head;
        ListNode slow = dummy;
        ListNode fast = head;

        while (fast != null && fast.next != null) {
            slow = slow.next;
            fast = fast.next.next;
            if (slow == fast) {
                return true;
            }
        }
        return false;
    }

    static class ListNode {
        int val;
        ListNode next;

        public ListNode(int val) {
            this.val = val;
        }
    }

}
